### ISSUE TITLE
F821 On class's type in type signature within the same class.

### ISSUE THREAD

**Issue by 0x6d6e647a** [astral-sh/ruff#6991]
I'm using Ruff v0.0.286 in the following.

For the following Python code:
```python
#!/usr/bin/env python

class Foo:
    def __init__(self, name: str) -> None:
        self._name = str

    @staticmethod
    def from_int(n: int) -> Foo:
        return Foo(str(n))

x = Foo.from_int(5)
``` 
Ruff generates the following error:
```shell
user@host $ ruff test.py
test.py:8:29: F821 Undefined name `Foo`
Found 1 error.
``` 

Whereas Mypy says there is no error:
```shell
user@host $ mypy test.py
Success: no issues found in 1 source file
```

It seems that Ruff is not aware of a class's type definition from within the same class but only in side a type signature.

Please note that Ruff **correctly** does not error on the next line `return Foo(str(n))`. The issue seems to only happen within type signatures.

**Comment by zanieb** [comment]
Hi!

I believe our behavior is correct here, the class is _not_ available in the signature scope:

```
❯ python example.py
Traceback (most recent call last):
  File "/Users/mz/eng/src/astral-sh/ruff/example.py", line 3, in <module>
    class Foo:
  File "/Users/mz/eng/src/astral-sh/ruff/example.py", line 8, in Foo
    def from_int(n: int) -> Foo:
                            ^^^
NameError: name 'Foo' is not defined
```

**Comment by 0x6d6e647a** [comment]
You're correct! I fixed the code as follows:

```python
#!/usr/bin/env python
from typing import Self

class Foo:
    def __init__(self, name: str) -> None:
        self._name = str

    @classmethod
    def from_int(cls, n: int) -> Self:
        return cls(str(n))

x = Foo.from_int(5)
```

Sorry for wasting your time. I really appreciate the quick reply! Thank you very much!

**Comment by zanieb** [comment]
No problem! You can also use `"Foo"` but `Self` is a great new feature.
